server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreSnapshotMetrics.java

```diff
@@ -0,0 +1,115 @@
+    private final CounterMetric uploadTimeInNanos = new CounterMetric();
+    private final CounterMetric uploadReadTimeInNanos = new CounterMetric();
+    private final CounterMetric numberOfShardSnapshotsStarted = new CounterMetric();
+    private final CounterMetric numberOfShardSnapshotsCompleted = new CounterMetric();
+    private final Map<String, Object> metricAttributes;
+
+    public BlobStoreSnapshotMetrics(@Nullable ProjectId projectId, RepositoryMetadata repositoryMetadata, SnapshotMetrics snapshotMetrics) {
+        if (projectId != null) {
+            this.snapshotMetrics = snapshotMetrics;
+            metricAttributes = SnapshotMetrics.createAttributesMap(projectId, repositoryMetadata);
+        } else {
+            // Project ID should only be null for the stateless main blobstore, which is not used for snapshots
+            this.snapshotMetrics = SnapshotMetrics.NOOP;
+            this.metricAttributes = Map.of();
+        }
+    }
+
+    public void incrementSnapshotRateLimitingTimeInNanos(long throttleTimeNanos) {
+        snapshotMetrics.createThrottleDurationCounter().incrementBy(throttleTimeNanos, metricAttributes);
+        snapshotRateLimitingTimeInNanos.inc(throttleTimeNanos);
+    }
+
+    public long snapshotRateLimitingTimeInNanos() {
+        return snapshotRateLimitingTimeInNanos.count();
+    }
+
+    public void incrementRestoreRateLimitingTimeInNanos(long throttleTimeNanos) {
+        snapshotMetrics.restoreThrottleDurationCounter().incrementBy(throttleTimeNanos, metricAttributes);
+        restoreRateLimitingTimeInNanos.inc(throttleTimeNanos);
+    }
+
+    public long restoreRateLimitingTimeInNanos() {
+        return restoreRateLimitingTimeInNanos.count();
+    }
+
+    public void incrementCountersForPartUpload(long partSizeInBytes, long partWriteTimeNanos) {
+        snapshotMetrics.bytesUploadedCounter().incrementBy(partSizeInBytes, metricAttributes);
+        snapshotMetrics.uploadDurationCounter().incrementBy(partWriteTimeNanos, metricAttributes);
+        numberOfBytesUploaded.inc(partSizeInBytes);
+        uploadTimeInNanos.inc(partWriteTimeNanos);
+    }
+
+    public void incrementNumberOfBlobsUploaded() {
+        snapshotMetrics.blobsUploadedCounter().incrementBy(1, metricAttributes);
+        numberOfBlobsUploaded.inc();
+    }
+
+    public void shardSnapshotStarted() {
+        snapshotMetrics.shardsStartedCounter().incrementBy(1, metricAttributes);
+        numberOfShardSnapshotsStarted.inc();
+        shardSnapshotsInProgress.inc();
+    }
+
+    public void shardSnapshotCompleted(IndexShardSnapshotStatus status) {
+        final Map<String, Object> attrsWithStage = Maps.copyMapWithAddedEntry(metricAttributes, "stage", status.getStage().name());
+        snapshotMetrics.shardsCompletedCounter().incrementBy(1, attrsWithStage);
+        snapshotMetrics.shardsDurationHistogram().record(status.getTotalTime() / 1_000f, attrsWithStage);
```

Comment:
Why a `float` (`1_000f`) here rather than a `double`? Also it seems we don't use a `DoubleHistogram` for tracking the distribution of durations anywhere else, instead we typically use a `LongHistogram` with the times in milliseconds.

Reply:
Fixed it to use double instead of float in 4a8fb9d31b17b6422172fe346877e3621aff7aee